packages/repository-json-schema/src/build-schema.ts

```diff
@@ -176,34 +160,40 @@ export function modelToJsonSchema(ctor: Function): JSONSchema {
     // populating "properties" key
     result.properties[p] = metaToJsonProperty(metaProperty);
 
+    // handling 'required' metadata
+    if (metaProperty.required) {
+      result.required = result.required || [];
+      result.required.push(p);
+    }
+
     // populating JSON Schema 'definitions'
-    const referenceType = isArrayType(metaProperty.type as string | Function)
+    // shimks: ugly type casting; this should be replaced by logic to throw
+    // error if itemType/type is not a string or a function
+    const resolvedType = resolveType(metaProperty.type) as string | Function;
+    const referenceType = isArrayType(resolvedType)
       ? // shimks: ugly type casting; this should be replaced by logic to throw
         // error if itemType/type is not a string or a function
-        (metaProperty.itemType as string | Function)
-      : (metaProperty.type as string | Function);
-    if (typeof referenceType === 'function' && isComplexType(referenceType)) {
-      const propSchema = getJsonSchema(referenceType);
-
-      if (propSchema && Object.keys(propSchema).length > 0) {
-        result.definitions = result.definitions || {};
-
-        // delete nested definition
-        if (propSchema.definitions) {
-          for (const key in propSchema.definitions) {
-            result.definitions[key] = propSchema.definitions[key];
-          }
-          delete propSchema.definitions;
-        }
+        resolveType(metaProperty.itemType as string | Function)
+      : resolvedType;
 
-        result.definitions[referenceType.name] = propSchema;
```

Comment:
I added a reverse check + `continue`, this part is best reviewed with white-space changes ignored. (See "Diff settings" at the top of the page.)